### PR TITLE
Add cascade sum support for Inductor CPP backend

### LINKED ISSUE #154703: [Inductor] Output discrepancy between Inductor and eager of mean with input of a large size tensor
### 🐛 Describe the bug

Here is the example code to reproduce:
```
import torch

size = (30000, 100000)

t = torch.rand(size, dtype=torch.float, device="cpu")
op = torch.mean
expected = op(t)
actual = torch.compile(op)(t)
print(f"Expected: {expected}, Actual: {actual}")
```
Run it with single thread to ensure a large enough size for each thread
```
numactl -C 0-0 -m 0 python test.py
```
Here is the output
```
Expected: 0.49999791383743286, Actual: 0.08947848528623581
```

### Versions

```
[conda] torch                     2.8.0a0+git3c54e56           dev_0    <develop>
[conda] torchao                   0.12.0+git080576bc           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
[conda] torchmetrics              1.0.3                    pypi_0    pypi
[conda] torchrec                  1.1.0a0+8211be7           dev_0    <develop>
[conda] torchtune                 0.6.0.dev20250124+cpu          pypi_0    pypi
[conda] torchvision               0.22.0a0+fab1188           dev_0    <develop>
[conda] triton                    3.2.0                    pypi_0    pypi

```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #154703

Add cascade summation support for Inductor CPP backend to improve precision for large size summation.

Currently, Inductor CPP directly do reduction for sum. As shown in #154703, when the size of the sum is large and the number of parallel is small, direct reduction will cause an intolerable precision loss:
```
extern "C"  void kernel(float* in_out_ptr0,
                       const float* in_ptr0)
{
    auto out_ptr0 = in_out_ptr0;
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(3000000000L); x0+=static_cast<int64_t>(16L))
            {
                {
                    if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(3000000000L)))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                        tmp_acc0_vec = tmp_acc0_vec + tmp0;
                    }
                }
            }
            tmp_acc0 = tmp_acc0 + at::vec::vec_reduce_all<float, 1>([](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>& y) { return x + y; }, tmp_acc0_vec);
            out_ptr0[static_cast<int64_t>(0L)] = static_cast<float>(tmp_acc0);
        }
    }
    {
        {
            {
                auto tmp0 = out_ptr0[static_cast<int64_t>(0L)];
                auto tmp1 = static_cast<float>(3000000000.0);
                auto tmp2 = tmp0 / tmp1;
                in_out_ptr0[static_cast<int64_t>(0L)] = tmp2;
            }
        }
    }
}
```

After adding cascade sum support:

```
extern "C"  void kernel(float* in_out_ptr0,
                       const float* in_ptr0)
{
    auto out_ptr0 = in_out_ptr0;
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            at::vec::Vectorized<float> masked_tmp_acc0_vec = at::vec::Vectorized<float>(0);
            CascadeSumHelper<float, 65536> scalar_cascade_helper0(static_cast<int64_t>(3000000000L));
            CascadeSumHelper<at::vec::Vectorized<float>, 65536> cascade_helper0(static_cast<int64_t>(187500000L));
            CascadeSumHelper<at::vec::Vectorized<float>, 65536> masked_cascade_helper0(static_cast<int64_t>(0L));
            for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(3000000000L); x0+=static_cast<int64_t>(16L))
            {
                {
                    if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(3000000000L)))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                        tmp_acc0_vec = cascade_sum_combine(tmp0, &cascade_helper0);
                    }
                }
            }
            tmp_acc0 = cascade_sum_final(&scalar_cascade_helper0);
            tmp_acc0_vec = cascade_sum_final(&cascade_helper0);
            masked_tmp_acc0_vec = cascade_sum_final(&masked_cascade_helper0);
            tmp_acc0 = tmp_acc0 + at::vec::vec_reduce_all<float, 1>([](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>& y) { return x + y; }, tmp_acc0_vec + masked_tmp_acc0_vec);
            out_ptr0[static_cast<int64_t>(0L)] = static_cast<float>(tmp_acc0);
        }
    }
    {
        {
            {
                auto tmp0 = out_ptr0[static_cast<int64_t>(0L)];
                auto tmp1 = static_cast<float>(3000000000.0);
                auto tmp2 = tmp0 / tmp1;
                in_out_ptr0[static_cast<int64_t>(0L)] = tmp2;
            }
        }
    }
}
```
This will inevitably reduce performance when cascade sum is turned on.
For the case shown in #154703: performance reduced by ~3%.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/csrc/inductor/cpp_prefix.h [hunk 1]
```diff
@@ -74,6 +74,77 @@ template <typename T, int N>
...
+          T(typename T::value_type(0)));
+    } else {
+      sum_stk.assign(std::max(depth, static_cast<uint64_t>(1)), T(0));
+    }
+  }
+};
+
+template <typename T, uint64_t kChunkSize = 0>
+inline T cascade_sum_combine(
+    T& acc,
```
Comment (leslie-fang-intel) [inline torch/csrc/inductor/cpp_prefix.h]: `acc` passed in but not used?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -4376,6 +4540,13 @@ def aggregate_reduction_prefix_suffix(outer_loop: "LoopLevel"):
                             replace_acc_name(
                                 tail_loop_kernel.reduction_suffix, name, new_name
                             )
+                        # If tail loop kernel is a scalar kernel, use cascade_sum_combine_direct instead of cascade_sum_combine
+                        # as the reduction vars are extended: tmp_acc -> tmp_acc_arr[].
+                        replace_acc_name(
+                            tail_loop_kernel.stores,
+                            "cascade_sum_combine",
+                            "cascade_sum_combine_direct",
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp.py]: Looks like we already have `cascade_sum_combine` for scalar kernel, can we use this one for `tail loop kernel` instead of the `cascade_sum_combine_direct`?

File: torch/csrc/inductor/cpp_prefix.h (hunk 1 above)
Comment (CaoE) [reply torch/csrc/inductor/cpp_prefix.h]: Thanks for your comments. Removed the unused acc.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (CaoE) [reply torch/_inductor/codegen/cpp.py]: Removed `cascade_sum_combine_direct` and used `acc = acc + value;` instead of `acc = cascade_sum_combine(value, ...);`

**Review by leslie-fang-intel — APPROVED** [review]

**Review by jansel — APPROVED** [review]
